swingset/src/main/java/com/nqadmin/swingset/SSDataNavigator.java

```diff
@@ -82,6 +84,26 @@
  * reverted using Undo button (has to be done manually by the user).
  */
 public class SSDataNavigator extends JPanel {
+	private static class RowSetState {
+		private boolean inserting;
+	}
+
+	// don't have to worry about concurrency, always EDT
+	private static Map<RowSet,RowSetState> rowSetState = new WeakHashMap<>();
```

Comment:
Brilliant!

Reply:
So one map for class SSDataNavigator containing parings of RowSet instances each with the RowSetState for that instance. Each RowSet should only have one navigator and vice versa. 1-1. This allows use of static throughout. If we have 20 screens the map would presumably have 20 entries?

Reply:
Yep, I was assuming 1-1. If there are 20 navigators, should be 20 entries.

I don't think this code cares if there is more than one navigator for a row set, a row set would still only have one state struct. (And if more than one, something else might break first.

Oh, hang on a second, the code pattern is
```
onInsertRow = false;
setInserting(rowSet, onInsertRow);
```
to make it easier to back out. and this pattern means multiple navigators for same row set, would get out of sync with the state. Now that there's by in for this technique, I'll change it to
```
setInserting(rowSet, false);
```
 and get rid of the `onInsertRow` variable.

Reply:
Considering creating an interface called `RowSetState` which gives read only access to the state for a given rowset, and then ...

I'll leave it as is, but if/when more state is required, can expose RowSetState.